cmd/raidz_test/raidz_test.c

```diff
@@ -0,0 +1,800 @@
+			}
+		}
+	}
+	return (err);
+}
+
+int
+run_rec_check(raidz_test_opts_t *opts)
+{
+	char **impl_name;
+	unsigned fn, err = 0;
+	zio_t *zio_test;
+	raidz_map_t *rm_test;
+
+	err = init_raidz_golden_map(opts, PARITY_PQR);
+	if (0 != err)
+		return (err);
+
+	IPRINT(DBLSEP);
+	IPRINT("Testing data reconstruction...\n");
+
+	for (impl_name = (char **)raidz_impl_names+1; *impl_name != NULL;
+	    impl_name++) {
+
+		IPRINT(SEP);
+		IPRINT("\tTesting [%s] implementation...", *impl_name);
+
+		if (0 != vdev_raidz_impl_set(*impl_name)) {
+			IPRINT("[SKIP]\n");
+			continue;
+		} else {
+			IPRINT("[SUPPORTED]\n");
+		}
+
+		/* create suitable raidz_map */
+		rm_test = init_raidz_map(opts, &zio_test, PARITY_PQR);
+		/* generate parity */
+		vdev_raidz_generate_parity(rm_test);
+
+		for (fn = 0; fn < RAIDZ_REC_NUM; fn++) {
+
+			IPRINT("\t\tTesting method [%s] ...",
+				raidz_rec_name[fn]);
+
+			if (0 == __run_rec_check(opts, rm_test, fn)) {
+				IPRINT("[PASS]\n");
+			} else {
+				err++;
+				IPRINT("[FAIL]\n");
+			}
+		}
+		/* tear down test raidz_map */
+		fini_raidz_map(&zio_test, &rm_test);
+	}
+
+	fini_raidz_map(&opts->zio_golden, &opts->rm_golden);
+
+	return (err);
+}
+
+static int
+run_test(raidz_test_opts_t *opts)
+{
+	int err = 0;
+
+	if (opts == NULL)
+		opts = &rto_opts;
+
+	print_opts(opts, B_FALSE);
+
+	err |= run_gen_check(opts);
+	err |= run_rec_check(opts);
+
+	return (err);
+}
+
+#define	SWEEP_RUNNING	0
+#define	SWEEP_FINISHED	1
+#define	SWEEP_ERROR	2
+#define	SWEEP_TIMEOUT	3
+
+static int sweep_state = 0;
+static raidz_test_opts_t failed_opts;
+
+static kmutex_t sem_mtx;
+static kcondvar_t sem_cv;
+static int max_free_slots;
+static int free_slots;
+
+static void *
+sweep_thread(void *arg)
+{
+	int err = 0;
+	raidz_test_opts_t *opts = (raidz_test_opts_t *) arg;
+	VERIFY(opts != NULL);
+
+	err = run_test(opts);
+
+	if (rto_opts.rto_sanity) {
+		/* 25% chance that a sweep test fails */
+		if (rand() < (RAND_MAX/4))
+			err = 1;
+	}
+
+	if (0 != err) {
+		mutex_enter(&sem_mtx);
+		memcpy(&failed_opts, opts, sizeof (raidz_test_opts_t));
+		sweep_state = SWEEP_ERROR;
+		mutex_exit(&sem_mtx);
+	}
+
+	umem_free(opts, sizeof (raidz_test_opts_t));
+
+	/* signal the next thread */
+	mutex_enter(&sem_mtx);
+	free_slots++;
+	cv_signal(&sem_cv);
+	mutex_exit(&sem_mtx);
+
+	thread_exit();
+
+	return (NULL);
+}
+
+#define	ARR_LEN(x)	(sizeof (x) / sizeof (0[x]))
+
+static int
+run_sweep(void)
+{
+	static const size_t dcols_v[] = { 1, 2, 3, 4, 5, 6, 7, 8 };
+	static const size_t ashift_v[] = { 9, 12 };
+	static const size_t offset_cnt = 4;
+	static const size_t size_v[] = { 1<<9, 21 * (1<<9), 13 * (1<<12), 1<<17,
+	    (1<<20) - (1<<12), SPA_MAXBLOCKSIZE };
+
+	size_t total_comb = ARR_LEN(size_v) * ARR_LEN(ashift_v) *
+	    ARR_LEN(dcols_v) * offset_cnt;
+	size_t tried_comb = 0;
+	hrtime_t time_diff, start_time = gethrtime();
+	raidz_test_opts_t *opts;
+	int a, d, o, s;
+
+	max_free_slots = free_slots = MAX(2, sysconf(_SC_NPROCESSORS_ONLN));
```

Comment:
FYI `boot_ncpus` could be used for `sysconf(_SC_NPROCESSORS_ONLN)` but I'm OK with either since this code is userspace only.